### PR TITLE
Add `revert-file` feature

### LINKED ISSUE #1901: Remove file from PR
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/refined-github/issues/1901)

Basically a duplicate of #1127.

Sometimes in a PR file changes slip in that you didn't intend. Maybe you didn't correctly stash, a commit hook made a formatting change, or some other thing happened. 

The ultimate result is that you want to undo changes to a particular file in your PR. It'd be nice to have a button on the header of every file in the diff to revert that change. 

Something like this...

image

It seems like this was added in #1398, but it's no longer present or working now. At least not in the FF version of the plugin. GitHub added the trashcan icon to actually _delete_ a file... heh, I almost did that >_>.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $80.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
### Submitted pull Requests
- [#2182 Add `revert-file` feature](https://issuehunt.io/r/sindresorhus/refined-github/pull/2182)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/refined-github/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by fregante**
Closes #1901

asdasd

## Test

- [x] Works for text files
- [x] Avoids binary files
- [x] Deletes new files
- [ ] Handle renamed files: Yes, but as 2 operations (delete new file, revert old)
- [x] Works for files in folders
- [x] Works for files with spaces
- [x] Works for initially-empty files
- [x] Works in repos without push access
- [x] Works when the base branch has changed the content of the same file
- [x] Works when the the PR contains merges or rebases
- [x] Works on private repos
- [x] Reverts file deletions
- [x] Works in Chrome and Firefox

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#1901: Remove file from PR](https://issuehunt.io/repos/51769689/issues/1901)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by bfred-it** [comment]
This seems to work correctly, except it doesn't work on repos you don't have push rights to 😬 which sucks.

A workaround is to use the form on the `/edit/` page to submit the changes **instead** of the API.

Alternatives? Am I doing anything wrong in the API? The PUT requests receives a 404 when trying to revert the first file on https://github.com/sindresorhus/refined-github/pull/2146/files

File: source/features/revert-file.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		const filePath = (menuItem.closest('[data-path]') as HTMLElement).dataset.path!;
+		const [, forkRepoURL, forkBranch]: string[] = /^([^:]+):(.+)$/.exec(select('.head-ref')!.title) || [];
+		const [originalFile, forkFile] = await Promise.all([
+			api.v3(`repos/${getRepoURL()}/contents/${filePath}?ref=${baseRefOid}`, {ignoreHTTPStatus: true}),
+			api.v3(`repos/${forkRepoURL}/contents/${filePath}?ref=${forkBranch}`, {ignoreHTTPStatus: true})
+		]);
+
+		if (originalFile.content === undefined) {
+			// The file was added by this PR. Click the "Delete file" link instead
+			(menuItem.nextElementSibling as HTMLElement).click();
```
Comment (zephraph) [inline source/features/revert-file.tsx]: Is there a bit of a safer way to do this than positionally?

**Comment by zephraph** [comment]
I'm gonna restate the problem just so I'm sure I really understand what's happening.

**Scenarios when it works**

1. When you revert a file from a PR that's on a branch in the current repo and you have push access to the repo
2. When you revert a file from a fork PR, but you have push access to the underlying fork

**Scenarios when it fails**
1. When you try to revert when you don't have push access (expected)
2. When you try to revert when you're granted upstream push access by the PR fork owner (for the scope of that PR), but you don't have push access to the whole fork repo. (unexpected)

The second failure case is what we're talking about, right?

**Comment by bfred-it** [comment]
Yeah I think the bug here is that the API doesn’t follow the “Allow changes from maintainers” checkbox. We had this issue on other features. 

Best case scenario is that someone already found a solution to this or I’m doing something wrong. Even if I report it to GitHub it’s not gonna be fixed for a while, I bet I’m not the first one to notice it. 

`isomorphic-git` requires cloning the whole repo, we can’t do that for a one-off operation.

File: source/features/revert-file.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+					baseRefOid
+				}
+			}
+		}`);
+
+		const filePath = (menuItem.closest('[data-path]') as HTMLElement).dataset.path!;
+		const [, forkRepoURL, forkBranch]: string[] = /^([^:]+):(.+)$/.exec(select('.head-ref')!.title) || [];
+		const [originalFile, forkFile] = await Promise.all([
+			api.v3(`repos/${getRepoURL()}/contents/${filePath}?ref=${baseRefOid}`, {ignoreHTTPStatus: true}),
+			api.v3(`repos/${forkRepoURL}/contents/${filePath}?ref=${forkBranch}`, {ignoreHTTPStatus: true})
```
Comment (bfred-it) [inline source/features/revert-file.tsx]: Here I’m using v3 to get the file because:

- I can’t find the API to save the file in v4, so I have to save with v3
- it’s not a good idea to GET with v4 and PUT with v3 because one returns text and the other wants base64: the maybe they treat data differently as well

If you know how to commit files with v4 let me know

**Comment by zephraph** [comment]
If you haven't filed an issue yet, I'd still do it. This is enough of a corner case that I doubt many folks have encountered it.

**Comment by bfred-it** [comment]
> If you haven't filed an issue yet, I'd still do it. This is enough of a corner case that I doubt many folks have encountered it.

I'm swamped at the moment, could you test and report that?

**Comment by bfred-it** [comment]
It looks like they know of the limitation and they don’t care. This thread doesn’t exist anymore: http://web.archive.org/web/20180813143533/https://platform.github.community/t/access-token-should-allow-push-if-allow-edits-from-maintainers-checked/1072

I’ll try to implement it by fetching and pushing GitHub.com pages instead of the API

**Comment by bfred-it** [comment]
This is ready now!

**Comment by sindresorhus** [comment]
If there's only one commit with one file in a PR, and I try to revert, it does nothing: https://github.com/sindresorhus/playground/pull/9 (Notice commit 2 and 3 being empty). Maybe we should just disable the "Revert file" feature when it cannot work?

**Comment by fregante** [comment]
The content of the commits is irrelevant, the feature just fetches the file content @ “the parent commit of the first commit” and tries to save its content by using the Edit field form. 

This is the exact situation where it should work. What happens when you click Revert file? The menu should not close and if there’s an error it should tell you to look in the console.

**Comment by fregante** [comment]
One thing this feature can’t undo at all is *meta* changes like chmod. The rest of them should be possible

File: source/features/revert-file.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	);
+}
+
+function init(): void {
+	delegate('#files', '.js-file-header-dropdown > summary', 'click', handleMenuOpening);
+}
+
+features.add({
+	id: __featureName__,
+	description: 'Revert all the changes to a file in a PR',
```
Comment (fregante) [inline source/features/revert-file.tsx]:
```suggestion
	description: 'Adds button to revert all the changes to a file in a PR.](',
```

**Comment by fregante** [comment]
Ok, this is good to go again, tested in both browsers.

**Comment by fregante** [comment]
I'm merging this
